// tailwindcss-classnames/es/generation/ConfigScanner.js
import { __assign } from "tslib";
import { defaultThemeConfig, defaultVariants } from './utils/defaultTailwindConfig';
import _ from 'lodash';
var ConfigScanner = /** @class */ (function () {
    // FIXME: theme config does not consider function values
    function ConfigScanner(tailwindConfig) {
        var _this = this;
        var _a;
        this.getPrefix = function () { return _this.prefix; };
        this.getSeparator = function () { return _this.separator; };
        this.getTheme = function () { return _this.themeConfig; };
        this.getVariants = function () { return _this.variantsConfig; };
        this.getThemeProperty = function (themeProperty) {
            return [
                Object.keys(_this.themeConfig[themeProperty]),
                Object.values(_this.themeConfig[themeProperty]),
            ];
        };
        this.prefix = _.isEmpty(tailwindConfig === null || tailwindConfig === void 0 ? void 0 : tailwindConfig.prefix) ? '' : tailwindConfig.prefix;
        this.separator = _.isEmpty(tailwindConfig.separator)
            ? ':'
            : tailwindConfig.separator;
        this.variantsConfig = _.isEmpty(tailwindConfig.variants)
            ? defaultVariants // Order does matter, defaultVariants will be overridden by themeVariants.
            : __assign(__assign({}, defaultVariants), tailwindConfig.variants);
        this.themeConfig = _.merge(__assign(__assign({}, defaultThemeConfig), tailwindConfig.theme), (_a = tailwindConfig.theme) === null || _a === void 0 ? void 0 : _a.extend);
    }
    return ConfigScanner;
}());
export { ConfigScanner };
