// tailwindcss-classnames/es/generation/default-classes/Flexbox.js
var flexDirection = ['flex-row', 'flex-row-reverse', 'flex-col', 'flex-col-reverse'];
var flexWrap = ['flex-no-wrap', 'flex-wrap', 'flex-wrap-reverse'];
var alignItems = ['items-stretch', 'items-start', 'items-center', 'items-end', 'items-baseline'];
var alignContent = [
    'content-start',
    'content-center',
    'content-end',
    'content-between',
    'content-around',
];
var alignSelf = ['self-auto', 'self-start', 'self-center', 'self-end', 'self-stretch'];
var justifyContent = [
    'justify-start',
    'justify-center',
    'justify-end',
    'justify-between',
    'justify-around',
];
var flex = ['flex-initial', 'flex-1', 'flex-auto', 'flex-none'];
var flexGrow = ['flex-grow', 'flex-grow-0'];
var flexShrink = ['flex-shrink', 'flex-shrink-0'];
var order = [
    'order-first',
    'order-last',
    'order-none',
    'order-1',
    'order-2',
    'order-3',
    'order-4',
    'order-5',
    'order-6',
    'order-7',
    'order-8',
    'order-9',
    'order-10',
    'order-11',
    'order-12',
];
export var FlexBox = {
    flexDirection: flexDirection,
    flexWrap: flexWrap,
    alignItems: alignItems,
    alignContent: alignContent,
    alignSelf: alignSelf,
    justifyContent: justifyContent,
    flex: flex,
    flexGrow: flexGrow,
    flexShrink: flexShrink,
    order: order,
};
