// tailwindcss-classnames/lib/generation/default-classes/all.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllClasses = exports.AllClassesFlat = void 0;
var tslib_1 = require("tslib");
var Accessibility_1 = require("./Accessibility");
var Backgrounds_1 = require("./Backgrounds");
var Borders_1 = require("./Borders");
var Effects_1 = require("./Effects");
var Flexbox_1 = require("./Flexbox");
var Grid_1 = require("./Grid");
var Interactivity_1 = require("./Interactivity");
var Layout_1 = require("./Layout");
var Sizing_1 = require("./Sizing");
var Spacing_1 = require("./Spacing");
var SVG_1 = require("./SVG");
var Tables_1 = require("./Tables");
var Transforms_1 = require("./Transforms");
var Transitions_1 = require("./Transitions");
var Typography_1 = require("./Typography");
exports.AllClassesFlat = tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, Accessibility_1.Accessibility), Backgrounds_1.Backgrounds), Borders_1.Borders), Effects_1.Effects), Flexbox_1.FlexBox), Grid_1.Grid), Interactivity_1.Interactivity), Layout_1.Layout), Sizing_1.Sizing), Spacing_1.Spacing), SVG_1.SVG), Tables_1.Tables), Transforms_1.Transforms), Transitions_1.Transitions), Typography_1.Typography);
exports.AllClasses = {
    Accessibility: Accessibility_1.Accessibility,
    Backgrounds: Backgrounds_1.Backgrounds,
    Borders: Borders_1.Borders,
    Effects: Effects_1.Effects,
    FlexBox: Flexbox_1.FlexBox,
    Grid: Grid_1.Grid,
    Interactivity: Interactivity_1.Interactivity,
    Layout: Layout_1.Layout,
    Sizing: Sizing_1.Sizing,
    Spacing: Spacing_1.Spacing,
    SVG: SVG_1.SVG,
    Tables: Tables_1.Tables,
    Transforms: Transforms_1.Transforms,
    Transitions: Transitions_1.Transitions,
    Typography: Typography_1.Typography,
};
