// tailwindcss-classnames/lib/generation/ConfigScanner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigScanner = void 0;
var tslib_1 = require("tslib");
var defaultTailwindConfig_1 = require("./utils/defaultTailwindConfig");
var lodash_1 = tslib_1.__importDefault(require("lodash"));
var ConfigScanner = /** @class */ (function () {
    // FIXME: theme config does not consider function values
    function ConfigScanner(tailwindConfig) {
        var _this = this;
        var _a;
        this.getPrefix = function () { return _this.prefix; };
        this.getSeparator = function () { return _this.separator; };
        this.getTheme = function () { return _this.themeConfig; };
        this.getVariants = function () { return _this.variantsConfig; };
        this.getThemeProperty = function (themeProperty) {
            return [
                Object.keys(_this.themeConfig[themeProperty]),
                Object.values(_this.themeConfig[themeProperty]),
            ];
        };
        this.prefix = lodash_1.default.isEmpty(tailwindConfig === null || tailwindConfig === void 0 ? void 0 : tailwindConfig.prefix) ? '' : tailwindConfig.prefix;
        this.separator = lodash_1.default.isEmpty(tailwindConfig.separator)
            ? ':'
            : tailwindConfig.separator;
        this.variantsConfig = lodash_1.default.isEmpty(tailwindConfig.variants)
            ? defaultTailwindConfig_1.defaultVariants // Order does matter, defaultVariants will be overridden by themeVariants.
            : tslib_1.__assign(tslib_1.__assign({}, defaultTailwindConfig_1.defaultVariants), tailwindConfig.variants);
        this.themeConfig = lodash_1.default.merge(tslib_1.__assign(tslib_1.__assign({}, defaultTailwindConfig_1.defaultThemeConfig), tailwindConfig.theme), (_a = tailwindConfig.theme) === null || _a === void 0 ? void 0 : _a.extend);
    }
    return ConfigScanner;
}());
exports.ConfigScanner = ConfigScanner;
